index.js

```diff
@@ -47,6 +50,56 @@ function fastifyWebsocket (fastify, opts, next) {
   const wss = new WebSocket.Server(wssOptions)
   fastify.decorate('websocketServer', wss)
 
+  async function injectWS (path = '/') {
+    const server2Client = new PassThrough()
+    const client2Server = new PassThrough()
+
+    const serverStream = new Duplexify(server2Client, client2Server)
+    const clientStream = new Duplexify(client2Server, server2Client)
+
+    const ws = new WebSocket(null, { isServer: false })
+    const head = Buffer.from([])
+
+    let resolve
+    const promise = new Promise(_resolve => { resolve = _resolve })
+
+    ws.on('open', () => {
+      clientStream.removeListener('data', onData)
+      resolve(ws)
+    })
+
+    const onData = (chunk) => {
+      // Assign the socket only if the upgrade was successful and the socket is open
+      // istanbul ignore next
+      if (chunk.toString().includes('HTTP/1.1 101 Switching Protocols')) {
+        ws._isServer = false
+        ws.setSocket(clientStream, head, { maxPayload: 0 })
+      }
+    }
+
+    clientStream.on('data', onData)
+
+    const req = {
+      method: 'GET',
+      headers: {
+        connection: 'upgrade',
+        upgrade: 'websocket',
+        'sec-websocket-version': 13,
+        'sec-websocket-key': randomBytes(16).toString('base64')
+      },
+      httpVersion: '1.1',
+      url: path,
+      [kWs]: serverStream,
+      [kWsHead]: head
+    }
+
+    websocketListenServer.emit('upgrade', req, req[kWs], req[kWsHead])
```

Comment:
If someone adds request hooks to their fastify instance that process the request before it gets upgraded, will this fire them? Deserves a test I think! See https://github.com/fastify/fastify-websocket#using-hooks

Reply:
Good catch, indeed is not firing. 
I think I should add another parameter in injectWS to cover this case.